apps/prairielearn/assets/scripts/instructorAssessmentInstancesClient.ts

```diff
@@ -0,0 +1,739 @@
+    return html`<small>${uniq.join(', ')}</small>`;
+  }
+
+  function timeRemainingLimitFormatter(value: string, row: AssessmentInstanceRow) {
+    return html`
+      ${value}
+      <span>
+        <button
+          class="btn btn-secondary btn-xs ml-1 time-limit-edit-button"
+          id="row${row.assessment_instance_id}PopoverTimeLimit"
+          data-row="${JSON.stringify(row)}"
+        >
+          <i class="bi-pencil-square" aria-hidden="true"></i>
+        </a>
+      </span>
+    `.toString();
+  }
+
+  function detailsLinkFormatter(value: string, row: AssessmentInstanceRow) {
+    const name = assessmentGroupWork ? row.group_name : row.uid;
+
+    let number;
+    if (!assessmentMultipleInstance) {
+      number = row.number === 1 ? '' : `#${row.number}`;
+    }
+    return `<a href="${urlPrefix}/assessment_instance/${value}">${assessmentSetAbbr}${assessmentNumber}${number} for ${name}</a>`;
+  }
+
+  function detailsLinkSorter(
+    valueA: number,
+    valueB: number,
+    rowA: AssessmentInstanceRow,
+    rowB: AssessmentInstanceRow,
+  ) {
+    let nameA: string | null, nameB: string | null, idA, idB;
+    if (assessmentGroupWork) {
+      (nameA = rowA.group_name), (nameB = rowB.group_name);
+      (idA = rowA.group_id ?? ''), (idB = rowB.group_id ?? '');
+    } else {
+      (nameA = rowA.uid), (nameB = rowB.uid);
+      (idA = rowA.user_id ?? ''), (idB = rowB.user_id ?? '');
+    }
+
+    // Compare first by UID/group name, then user/group ID, then
+    // instance number, then by instance ID.
+    let compare = nameA?.localeCompare(nameB ?? '');
+    if (!compare) compare = (parseInt(idA) ?? 0) - (parseInt(idB) ?? 0);
+    if (!compare) compare = (rowA.number ?? 0) - (rowB.number ?? 0);
+    if (!compare) compare = valueA - valueB;
+    return compare;
+  }
+
+  function timeRemainingLimitSorter(
+    valueA: number,
+    valueB: number,
+    rowA: AssessmentInstanceRow,
+    rowB: AssessmentInstanceRow,
+  ) {
+    // Closed assessments are listed first, followed by time limits
+    // ascending, followed by open without a time limit
+    return Number(rowA.open) - Number(rowB.open) || (valueA ?? Infinity) - (valueB ?? Infinity);
+  }
+
+  function actionButtonFormatter(_value: string, row: AssessmentInstanceRow) {
+    const ai_id: number = parseInt(row.assessment_instance_id);
+    return html`
+      <div>
+        <div class="dropdown">
+          <button
+            type="button"
+            class="btn btn-secondary btn-xs dropdown-toggle"
+            data-toggle="dropdown"
+            aria-haspopup="true"
+            aria-expanded="false"
+            data-boundary="window"
+          >
+            Action
+          </button>
+          <div
+            id="row${ai_id}PopoverClose"
+            tabindex="0"
+            data-toggle="popover"
+            title="Confirm close"
+            data-content="${escapeHtml(CloseForm({ csrfToken: csrfToken ?? '', ai_id }))}"
```

Comment:
If `csrfToken` is undefined, we shouldn't blindly continue to set an empty string here. Can we validate that `csrfToken` is defined earlier on?

Reply:
Now checking for `csrfToken` earlier and throwing an error if it is not defined. Is this what we want to be doing here? I figure we want to fail loudly if we are unable to submit but not sure on this one.

Reply:
Yep, we want to fail loudly if something isn't what we expect. Typically, this is how I'd think about things:

```
// OK:
if (value) {
  doThing(value);
}

// Preferred:
if (!value) throw new Error('Relevant error message goes here');
doThing(value);
```

Reply:
Sounds good, thanks!